shared/src/main/scala/kinesis4cats/producer/fs2/FS2Producer.scala

```diff
@@ -111,33 +112,36 @@ abstract class FS2Producer[F[_], PutReq, PutRes](implicit
         .evalMap { x =>
           val c = ctx.addEncoded("batchSize", x.size)
           x.toNel
-            .fold(F.unit) { records =>
+            .fold(F.unit) { x =>
+              val records = x.map(_._1)
+              val deferreds = x.map(_._2)
               for {
                 _ <- logger.debug(c.context)(
                   "Received batch to process"
                 )
-                _ <- underlying
-                  .put(records)
-                  .flatMap(callback)
+                _ <- F.guaranteeCase(
+                  underlying
+                    .put(records)
+                    .flatMap(callback)
+                ) {
+                  case Canceled() | Succeeded(_) =>
+                    deferreds.traverse_(_.complete(F.unit))
+                  case Errored(e) =>
+                    deferreds.traverse_(d => d.complete(F.raiseError(e)))
```

Comment:
**suggestion:** micro-optimization to avoid creating N effects with the same failure

```suggestion
                   val failedF = F.raiseError(e)
                    deferreds.traverse_(d => d.complete(failedF))
```

Reply:
I'm also wondering if this `guaranteeCase` should include the `callback` or not, but I'm not sure what `callback` even does.

Reply:
Callback is a configurable F that a user can provide against successful records. That is not really needed anymore I suppose.